servicetalk-http-api/src/main/java/io/servicetalk/http/api/AbstractZipContentCodec.java

```diff
@@ -0,0 +1,558 @@
+                    src.readBytes(onHeap);
+                    output.write(onHeap);
+                }
+            }
+
+            output.finish();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        } finally {
+            closeQuietly(output);
+        }
+
+        return dst;
+    }
+
+    @Override
+    public final Publisher<Buffer> encode(final Publisher<Buffer> from,
+                                          final BufferAllocator allocator) {
+        return from
+                .map((it) -> (Object) it)
+                .concat(succeeded(END_OF_STREAM))
+                .liftSync(subscriber -> new PublisherSource.Subscriber<Object>() {
+
+                    @Nullable
+                    Buffer dst;
+                    @Nullable
+                    DeflaterOutputStream output;
+
+                    @Override
+                    public void onSubscribe(PublisherSource.Subscription subscription) {
+                        try {
+                            dst = allocator.newBuffer(chunkSize);
+                            output = newDeflaterOutputStream(asOutputStream(dst));
+                        } catch (IOException e) {
+                            throw new RuntimeException(e);
+                        }
+                        subscriber.onSubscribe(subscription);
+                    }
+
+                    @Override
+                    public void onNext(Object next) {
+                        assert output != null;
+                        assert dst != null;
+
+                        // onNext will produce AT-MOST N items (as received)
+                        // +1 for the encoding footer (ie. END_OF_STREAM)
+                        try {
+                            if (next == END_OF_STREAM) {
+                                try {
+                                    output.finish();
+                                    subscriber.onNext(dst.readSlice(dst.readableBytes()));
+                                } catch (IOException e) {
+                                    throw new RuntimeException(e);
+                                }
+                                return;
+                            }
+
+                            Buffer src = (Buffer) next;
+                            if (src.hasArray()) {
+                                output.write(src.array(), src.readerIndex(), src.readableBytes());
+                            } else {
+                                while (src.readableBytes() > 0) {
+                                    byte[] onHeap = new byte[min(src.readableBytes(), chunkSize)];
+                                    src.readBytes(onHeap);
+                                    output.write(onHeap);
+                                }
+                            }
+
+                            output.flush();
+                            subscriber.onNext(dst.readSlice(dst.readableBytes()));
```

Comment:
The problem with `readSlice` here is that we keep writing new data to the original `dst` buffer. As the result, it will be resized multiple times and will allocate all the streaming data in memory 😢

Reply:
Good catch, lets keep this out of the scope, since we are discussing about the alternative implementation with `EmbeddedChannel`. I will keep track of that.

Reply:
Agreed that it's not worth spending cycles on fixing implementation that is going to be replaced soon. But this is an important aspect to consider before making a decision to include these implementations in the next release. Or we should clarify in the release notes that compression should not be used with streaming clients for now.

//cc @Scottmitch